src/algebra/big_operators/finprod.lean

```diff
@@ -0,0 +1,623 @@
+
+/-- Given finite sets `s` and `t`, the product of `f i` over `i ∈ s ∪ t` times the product of
+`f i` over `i ∈ s ∩ t` equals the product of `f i` over `i ∈ s` times the product of `f i`
+over `i ∈ t`. -/
+@[to_additive] lemma finprod_mem_union_inter (hs : s.finite) (ht : t.finite) :
+  (∏ᶠ i ∈ s ∪ t, f i) * ∏ᶠ i ∈ s ∩ t, f i = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ t, f i :=
+begin
+  unfreezingI { lift s to finset α using hs, lift t to finset α using ht },
+  rw [← finset.coe_union, ← finset.coe_inter],
+  simp only [finprod_mem_coe_finset, finset.prod_union_inter]
+end
+
+/-- A more general version of `finprod_mem_union_inter` that requires `s ∩ mul_support f` and
+`t ∩ mul_support f` instead of `s` and `t` to be finite. -/
+@[to_additive] lemma finprod_mem_union_inter'
+  (hs : (s ∩ mul_support f).finite) (ht : (t ∩ mul_support f).finite) :
+  (∏ᶠ i ∈ s ∪ t, f i) * ∏ᶠ i ∈ s ∩ t, f i = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ t, f i :=
+begin
+  rw [← finprod_mem_inter_mul_support f s, ← finprod_mem_inter_mul_support f t,
+    ← finprod_mem_union_inter hs ht, ← union_inter_distrib_right, finprod_mem_inter_mul_support,
+    ← finprod_mem_inter_mul_support f (s ∩ t)],
+  congr' 2,
+  rw [inter_left_comm, inter_assoc, inter_assoc, inter_self, inter_left_comm]
+end
+
+/-- A more general version of `finprod_mem_union` that requires `s ∩ mul_support f` and
+`t ∩ mul_support f` instead of `s` and `t` to be finite. -/
+@[to_additive] lemma finprod_mem_union' (hs : (s ∩ mul_support f).finite)
+  (ht : (t ∩ mul_support f).finite) (hst : disjoint s t) :
+  ∏ᶠ i ∈ s ∪ t, f i = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ t, f i :=
+by rw [← finprod_mem_union_inter' hs ht, disjoint_iff_inter_eq_empty.1 hst, finprod_mem_empty,
+  mul_one]
+
+/-- Given two finite disjoint sets `s` and `t`, the product of `f i` over `i ∈ s ∪ t` equals the
+product of `f i` over `i ∈ s` times the product of `f i` over `i ∈ t`. -/
+@[to_additive] lemma finprod_mem_union (hs : s.finite) (ht : t.finite) (hst : disjoint s t) :
+  ∏ᶠ i ∈ s ∪ t, f i = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ t, f i :=
+finprod_mem_union' (hs.inter_of_left _) (ht.inter_of_left _) hst
+
+/-- A more general version of `finprod_mem_union'` that requires `s ∩ mul_support f` and
+`t ∩ mul_support f` instead of `s` and `t` to be disjoint -/
+@[to_additive] lemma finprod_mem_union'' (hs : (s ∩ mul_support f).finite)
+  (ht : (t ∩ mul_support f).finite) (hst : disjoint (s ∩ mul_support f) (t ∩ mul_support f)) :
+  ∏ᶠ i ∈ s ∪ t, f i = (∏ᶠ i ∈ s, f i) * ∏ᶠ i ∈ t, f i :=
+by rw [← finprod_mem_inter_mul_support f s, ← finprod_mem_inter_mul_support f t,
+  ← finprod_mem_union hs ht hst, ← union_inter_distrib_right, finprod_mem_inter_mul_support]
+
+/-- The product of `f i` over `i ∈ {a}` equals `f a`. -/
+@[to_additive] lemma finprod_mem_singleton : ∏ᶠ i ∈ ({a} : set α), f i = f a :=
+by rw [← finset.coe_singleton, finprod_mem_coe_finset, finset.prod_singleton]
+
+@[simp, to_additive] lemma finprod_finprod_eq_left : ∏ᶠ i = a, f i = f a :=
```

Comment:
This lemma could use a better name (but we shouldn't use `mem` because there is no `∈`).